binary_search_tree/tree.py

```diff
@@ -1,59 +1,117 @@
 class TreeNode:
-    def __init__(self, key, val = None):
-        if val == None:
-            val = key
+    def __init__(self, key, value=None):
+        if value == None:
+            value = key
 
         self.key = key
-        self.value = val
+        self.value = value
         self.left = None
         self.right = None
-        
 
 
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
-    def add(self, key, value = None):
-        pass
+    # Time Complexity: O(log(n))*assumes all balanced tree
+    # Space Complexity: O(1)
+    def add(self, key, value=None):
+        if self.root == None:
+            self.root = TreeNode(key, value)
+        else:
+            self.add_helper(self.root, key, value)
 
-    # Time Complexity: 
-    # Space Complexity: 
-    def find(self, key):
-        pass
+    def add_helper(self, current_node, key, value):
+        if current_node == None:
+            return TreeNode(key, value)
+        if key < current_node.key:
+            current_node.left = self.add_helper(current_node.left, key, value)
+        else:
+            current_node.right = self.add_helper(current_node.right, key, value)
+        return current_node
+
+    # Time Complexity: O(log(n))
+    # Space Complexity: O(1)
+    def find(self, value):
+        return self.find_helper(self.root, value)
+
+    def find_helper(self, current, key):
+        if current == None:
+            return None
+        elif current.key == key:
+            return current.value
+        elif key < current.key:
+            return self.find_helper(current.left, key)
+        return self.find_helper(current.right, key)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(logn)
+    # Space Complexity: O(n)
     def inorder(self):
```

Comment:
👍  Since you hit each node and make a list the of all the nodes in the tree the time and space complexity is O(n)